# Pandas lab 1

## **Import data and libraries**

In this Lab we will use a restaurant dataset. 

The dataset contains tips data from different customers females and males, smokers and non smokers, from days, dinner or lunch and from different table size.

In [3]:
import pandas as pd

# Load restaurant.csv file from data folder to variable "df":
df = pd.read_csv('./data/restaurant.csv')


In [4]:
# show the first 5 rows
# your code here
df.head(5)

total_bill   tip     sex smoker  day    time  size
0       16.99  1.01  Female     No  Sun  Dinner     2
1       10.34  1.66    Male     No  Sun  Dinner     3
2       21.01  3.50    Male     No  Sun  Dinner     3
3       23.68  3.31    Male     No  Sun  Dinner     2
4       24.59  3.61  Female     No  Sun  Dinner     4

## **Overview**

### 📑 Question 1 (EXPRESSION - 5 pts)

Write **one line of code** to Show the **first 3 rows** of the data

**Output should look like this**

|    |   total_bill |   tip | sex    | smoker   | day   | time   |   size |
|---:|-------------:|------:|:-------|:---------|:------|:-------|-------:|
|  0 |        16.99 |  1.01 | Female | No       | Sun   | Dinner |      2 |
|  1 |        10.34 |  1.66 | Male   | No       | Sun   | Dinner |      3 |
|  2 |        21.01 |  3.5  | Male   | No       | Sun   | Dinner |      3 |

In [2]:
# Your answer/code here:


### 📑 Question 2 (EXPRESSION - 5 pts)

Write **one line of code** to Show the **last 2 rows** of the data

**Output should look like this**

|     |   total_bill |   tip | sex    | smoker   | day   | time   |   size |
|----:|-------------:|------:|:-------|:---------|:------|:-------|-------:|
| 242 |        17.82 |  1.75 | Male   | No       | Sat   | Dinner |      2 |
| 243 |        18.78 |  3    | Female | No       | Thur  | Dinner |      2 |

In [3]:
# Your answer/code here:


### 📑 Question 3 (MULTICHOICE_SINGLE - 5 pts)

What syntax is right to get random rows in dataframe?

-  A df.shuffle()

- B df.random()

- C df.sample()

- D df.mix()

In [4]:
# Your answer/code here:


### 📑 Question 4 (EXPRESSION - 5 pts)
Write one line of code to show a **descriptive statistics summary** of the dataset.

**Output should look like this**

|       |   total_bill |       tip |      size |
|:------|-------------:|----------:|----------:|
| count |    244       | 244       | 244       |
| mean  |     19.7859  |   2.99828 |   2.56967 |
| std   |      8.90241 |   1.38364 |   0.9511  |
| min   |      3.07    |   1       |   1       |
| 25%   |     13.3475  |   2       |   2       |
| 50%   |     17.795   |   2.9     |   2       |
| 75%   |     24.1275  |   3.5625  |   3       |
| max   |     50.81    |  10       |   6       |

In [5]:
# Your answer/code here:


## **EDA - Exploratory Data Analysis**

### 📑 Question 5 (EXPRESSION - 5 pts)
Write one line of code to find out the **minimum amount of tip**.

**Output should be a number**

In [8]:
# Your answer/code here:


### 📑 Question 6 (EXPRESSION - 5 pts)
Write one line of code to find out the **maximum amount of tip**.

**Output should be a number**

In [9]:
# Your answer/code here:


### 📑 Question 7 (MULTICHOICE_SINGLE - 5 pts)

On which days of the week, the restaurant is open?

- A. Mon, Tue, Wed, Thu

- B. Thu, Fri, Sat, Sun

- C. Mon, Wed, Fri 

- D. Open all days in week

In [2]:
# Your answer/code here:


array(['Sun', 'Sat', 'Thur', 'Fri'], dtype=object)

### 📑 Question 8 (EXPRESSION - 5 pts)
Write one line of code to get the list of **unique values** of the column ```time```.

**Output should look like this**

array(['Dinner', 'Lunch'], dtype=object)

In [11]:
# Your answer/code here:


### 📑 Question 9 (EXPRESSION - 5 pts)
Write one line of code to get the range of table size. 

Range means the **difference between the max and the min**.

**Output should be a number**

In [12]:
# Your answer/code here:


### 📑 Question 10 (MULTICHOICE_SINGLE - 5 pts)
How many men/women paid for the dish?

- A Male: 157, Female: 89

- B Male: 85, Female: 156

- C Male: 157, Female: 87

- D Male: 142, Female: 15

In [13]:
# Your answer/code here:


### 📑 Question 11 (MULTICHOICE_SINGLE - 5 pts)
What is the percentage of number of smoker table over the total?

- A 38%

- B 40%

- C 2%

- D 62%

In [14]:
# Your answer/code here:


### 📑 Question 12 (MULTICHOICE_SINGLE - 5 pts)
What is the percentage of revenue from the smoker tables over the total?

- A 38%

- B 40%

- C 2%

- D 62%

In [15]:
# Your answer/code here:


### 📑 Question 14 (MULTICHOICE_SINGLE - 5 pts)
Between smoker and non-smoker, who tips more according the percentage of the total tip on total bill.

- A Non-smoker

- B Smoker

In [17]:
# Your answer/code here:


### 📑 Question 15 (EXPRESSION - 5 pts)

Write one line of code to filter out the **"smoker"** tables on **Sunday**.

**Output should look like this**

|     |   total_bill |   tip | sex    | smoker   | day   | time   |   size |
|----:|-------------:|------:|:-------|:---------|:------|:-------|-------:|
| 164 |        17.51 |  3    | Female | Yes      | Sun   | Dinner |      2 |
| 172 |         7.25 |  5.15 | Male   | Yes      | Sun   | Dinner |      2 |
| 173 |        31.85 |  3.18 | Male   | Yes      | Sun   | Dinner |      2 |
| 174 |        16.82 |  4    | Male   | Yes      | Sun   | Dinner |      2 |
| 175 |        32.9  |  3.11 | Male   | Yes      | Sun   | Dinner |      2 |
| 176 |        17.89 |  2    | Male   | Yes      | Sun   | Dinner |      2 |
| 177 |        14.48 |  2    | Male   | Yes      | Sun   | Dinner |      2 |
| 178 |         9.6  |  4    | Female | Yes      | Sun   | Dinner |      2 |
| 179 |        34.63 |  3.55 | Male   | Yes      | Sun   | Dinner |      2 |
| 180 |        34.65 |  3.68 | Male   | Yes      | Sun   | Dinner |      4 |
| 181 |        23.33 |  5.65 | Male   | Yes      | Sun   | Dinner |      2 |
| 182 |        45.35 |  3.5  | Male   | Yes      | Sun   | Dinner |      3 |
| 183 |        23.17 |  6.5  | Male   | Yes      | Sun   | Dinner |      4 |
| 184 |        40.55 |  3    | Male   | Yes      | Sun   | Dinner |      2 |
| 186 |        20.9  |  3.5  | Female | Yes      | Sun   | Dinner |      3 |
| 187 |        30.46 |  2    | Male   | Yes      | Sun   | Dinner |      5 |
| 188 |        18.15 |  3.5  | Female | Yes      | Sun   | Dinner |      3 |
| 189 |        23.1  |  4    | Male   | Yes      | Sun   | Dinner |      3 |
| 190 |        15.69 |  1.5  | Male   | Yes      | Sun   | Dinner |      2 |

In [18]:
# Your answer/code here:


### 📑 Question 16 (MULTICHOICE_SINGLE - 5 pts)
How many tables belong to the smoker, at lunch?

- A 20

- B 21 

- C 22

- D 23

In [19]:
# Your answer/code here:


### 📑 Question 17 (EXPRESSION - 10 pts)

Write one line of code to compute the **average tip** amount of men who smoke and dine on **Friday's dinner**.


**Output should be a number**

In [20]:
# Your answer/code here:


### 📑 Question 18 (EXPRESSION - 5 pts)
Write one line of code to compute the **sum** of total bills of the **male** customers who dine in any day that's **not Friday**. 

**Output should be a number**

In [21]:
# Your answer/code here:


### 📑 Question 19 (EXPRESSION - 5 pts)
Write one line of code to filter the tables with tip amount **larger** than the **average** tip amount. 

In [22]:
# Your answer/code here:


### 📑 Question 20 (EXPRESSION - 5 pts)

- Task 1: load `time_hours.xlsx` table to `df_th` variable:

In [ ]:
df_th = _ # your code here

# show time_hours table:
df_th

time    hour
0  Dinner  19-21h
1   Lunch  10-12h

- Task 2: Get `hour` information for `df` dataframe from `df_th` by using Using `time` field (`pd.merge()`):

In [ ]:
# your code here:


In [ ]:
# show random sample 5 rows:
df.sample(5)

total_bill   tip     sex smoker   day    time  size    hour
234       15.53  3.00    Male    Yes   Sat  Dinner     2  19-21h
22        15.77  2.23  Female     No   Sat  Dinner     2  19-21h
124       12.48  2.52  Female     No  Thur   Lunch     2  10-12h
81        16.66  3.40    Male     No  Thur   Lunch     2  10-12h
147       11.87  1.63  Female     No  Thur   Lunch     2  10-12h